rep-2011.rst

```diff
@@ -0,0 +1,1314 @@
+                          └───────────────────────┘    │
+                                                       │
+                           possibly other transfer functions
+
+The tool will start with the current type version and see if it can be automatically converted to the desired type version, or if it is accepted as an input to any user-defined transfer functions or if it can be automatically converted into one of the input type versions for the transfer functions.
+It will continue to do this until it reaches the desired type version or it fails to find a path from the current to the desired type version.
+
+.. code::
+
+    ┌──────────────────────┐      /topic      ┌─────────────────────────┐
+    │Publisher<Message@ABC>├────────X────────►│Subscription<Message@XYZ>│
+    └──────────────────────┘                  └─────────────────────────┘
+                                  │ │ │
+                 remap publisher  │ │ │  and add transfer function
+                                  ▼ ▼ ▼
+    ┌──────────────────────┐                  ┌─────────────────────────┐
+    │Publisher<Message@ABC>│                  │Subscription<Message@XYZ>│
+    └─┬────────────────────┘                  └─────────────────────────┘
+      │                                                          ▲
+      │             ┌─────────────────────────────────┐          │
+      ✓ /topic/ABC  │ Transfer Functions for ABC->XYZ │   /topic ✓
+      │             │                                 │          │
+      │  ┌──────────┴──────────────┐   ┌──────────────┴───────┐  │
+      └─►│Subscription<Message@ABC>│   │Publisher<Message@XYZ>├──┘
+         └──────────┬──────────────┘   └──────────────┬───────┘
+                    │                                 │
+                    └─────────────────────────────────┘
+
+Once the set of necessary transfer functions has been identified, the ROS graph can be changed to have one side of the topic be remapped onto a new topic name which indicates it is of a different version that what is desired, and then the transfer function can be run as a component node which subscribes to one version of the message, performs the conversion using the chain of transfer functions, and then publishes the other version of the message.
+Tools will assist the user in making these remappings and running the necessary component nodes with the appropriate configurations, either from their launch file or from the command line.
+
+.. TODO::
+
+    discuss the implications for large messages and the possibility of having the transfer functions be colocated with either the publisher or subscription more directly than with component nodes and remapping.
+
+Once the mismatched messages are flowing through the transfer functions, communication should be possible and neither the publishing side nor the subscribing side have any specific knowledge of the conversions taking place or that any conversions are necessary.
+
```

Comment:
Transfer functions would be spun up between mismatched publishers and subscribers, so I guess the only way for an attacker to scale the attack would be to spin up a whole bunch of subscribers that aren't well matched with a publisher (scaling the attack laterally.) I don't see how it would be possible to have a big linear chain of conversions in this case.

That said, I suppose it would be possible to have a whitelisting feature in the type description distribution section, or on the transfer functions. Something to think about.

Reply:
Yeah, in my mind the transfer functions would not be started automatically. You'd have to opt-in to running them (add something to your launch file or otherwise to your application) or even if we came up with something that would start them automatically the list of transfer functions would have to be defined ahead of time (as they must be local to the machine running the conversion) and we will already have to assert that there are no cycles in the proposed chain of transfer functions. So I don't see a vector for an attacker to use in this part of the system right now, but I could be wrong.

I didn't add a security section because the conversions happen over topics and services as any application would, and so use of transfer functions would require the same security considerations as using topics between any other two parts of your system. I.e. the transfer functions aren't special, they're just another node that uses topics.

Reply:
Thanks both for the answers.

> Yeah, in my mind the transfer functions would not be started automatically. 

I think this approach to would do. Nevertheless, I'd suggest that once implemented these possible issues get (pen)tested.

> I didn't add a security section because the conversions happen over topics and services as any application would, and so use of transfer functions would require the same security considerations as using topics between any other two parts of your system. I.e. the transfer functions aren't special, they're just another node that uses topics.

Then, maybe mention this explicitly? Security isn't mentioned in the document even once right now. Also, though I agree that the (security) mechanisms might be same, it's not easy currently to properly secure ROS 2 graphs. We confirmed this with https://arxiv.org/pdf/2208.02615.pdf. When attempted with other graphs (e.g. MoveIt2)  results indicated that things start becoming really complex and cumbersome. To that we'd add the added complexity introduced by features in here. 

Just a thought, I believe that the thought exercise now providing examples/guidance on security would help significantly in the future.

Reply:
You're right, it's worth mentioning, even if to say "they're just topics/services and are handled like any other ones" and to mention the addition of the type version hash to the discovery information.

@methylDragon do you want to take a stab at adding that section?

Reply:
Probably its own section directly under the `Specification` section (at the end, after the other sections, as they're needed for context), or maybe even as a peer of that section, directly after it, before the the rationale section. I'm not 100% sure though.